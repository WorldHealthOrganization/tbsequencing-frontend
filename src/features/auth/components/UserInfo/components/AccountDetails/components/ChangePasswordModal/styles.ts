import { css } from '@emotion/react';

export const buttonsContainer = css({
  display: 'flex',
  marginTop: 32,
  justifyContent: 'flex-end',
});

export const fieldWrapper = css({
  marginTop: 16,
});

export const changePasswordBtn = css({
  marginLeft: 16,
});

export const fieldsContainer = css({
  marginTop: 32,
  marginBottom: 32,
});
