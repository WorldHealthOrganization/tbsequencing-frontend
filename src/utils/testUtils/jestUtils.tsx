import React, { ReactElement } from 'react';
import { render as rtlRender } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import { reducer } from '../../app/store';

const getWrapper = (
  ui: ReactElement,
  {
    // @ts-ignore
    preloadedState,
    store = configureStore({ reducer, preloadedState }),
  } = {},
) => ({ children }: { children: ReactElement }) => (
  <Provider store={store}>{children}</Provider>
);

function render(
  ui: ReactElement,
  {
    // @ts-ignore
    preloadedState,
    store = configureStore({ reducer, preloadedState }),
    ...renderOptions
  } = {},
) {
  const Wrapper = getWrapper(ui, { preloadedState, store });

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

const getTreeJSONWithRedux = (ui: ReactElement, {
  // @ts-ignore
  preloadedState,
  store = configureStore({ reducer, preloadedState }),
} = {}) => {
  const Wrapper = getWrapper(ui, { preloadedState, store });

  const tree = renderer
    .create(<Wrapper>{ui}</Wrapper>)
    .toJSON();

  return tree;
};

export const wrappedInRouter = (
  component: ReactElement,
) => <BrowserRouter>{component}</BrowserRouter>;

export * from '@testing-library/react';
export { render, getTreeJSONWithRedux };
