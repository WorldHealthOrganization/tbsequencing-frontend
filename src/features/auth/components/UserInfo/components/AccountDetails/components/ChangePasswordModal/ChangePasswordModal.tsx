/** @jsxImportSource @emotion/react */
import React, { useEffect } from 'react';
import { Control, FieldErrors, useForm } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers/yup';
import * as yup from 'yup';
import { toast } from 'react-toastify';
import * as styles from './styles';
import BaseModal from '../../../../../../../../components/BaseModal';
import H1 from '../../../../../../../../components/typography/H1';
import { AbstractIField } from '../../../../../registration/types';
import FormFields from '../../../../../../../../components/FormFields';
import { validatePassword } from '../../../../../../../../utils/validators';
import { AppButton } from '../../../../../../../../components/AppButton/AppButton';
import { useChangePasswordMutation } from '../../../../../../../../services/indentityApi/identityApi';
import { IHttpError } from '../../../../../../../../models/HttpError';
import { useHttpError } from '../../../../../../../../hooks/useHttpError';

interface Props {
  isOpened: boolean;
  handleClose: () => void;
}

interface IInputs {
  oldPassword: string;
  newPassword1: string;
  newPassword2: string;
}

const defaultValues: IInputs = {
  oldPassword: '',
  newPassword1: '',
  newPassword2: '',
};

const formFields:AbstractIField<IInputs>[] = [
  {
    name: 'oldPassword',
    type: 'password',
    placeholder: 'Enter your password',
    label: 'Old Password',
  },
  {
    name: 'newPassword1',
    type: 'password',
    placeholder: 'Enter your password',
    deps: 'newPassword2',
    showPasswordHint: true,
    label: 'New Password',
    wrapperStyles: styles.fieldWrapper,
  },
  {
    name: 'newPassword2',
    type: 'password',
    deps: 'newPassword1',
    placeholder: 'Repeat password',
    label: 'Confirm Password',
    wrapperStyles: styles.fieldWrapper,
  },
];

const formSchema = yup.object().shape({
  newPassword1: validatePassword,
  newPassword2: yup.string().oneOf([yup.ref('newPassword1')], 'Passwords do not match'),
  oldPassword: validatePassword,
});

export const ChangePasswordModal = ({ isOpened, handleClose }: Props) => {
  const {
    control, formState: { errors, isValid }, setError, handleSubmit, reset,
  } = useForm<IInputs>({
    resolver: yupResolver(formSchema),
    defaultValues,
    mode: 'onChange',
  });

  const [changePassword, { error, data }] = useChangePasswordMutation();

  useHttpError<IInputs>(error as IHttpError, setError);

  useEffect(() => {
    if (data) {
      toast.success('Password was successfully changed');
      reset();
      handleClose();
    }
  }, [data]);

  const onSubmit = (fieldData: IInputs) => {
    changePassword(fieldData);
  };

  const onCancel = () => {
    handleClose();
    reset(defaultValues);
  };

  return (
    <BaseModal open={isOpened}>
      <div>
        <H1>Change password</H1>
        <div css={styles.fieldsContainer}>
          <FormFields<IInputs>
            control={control as unknown as Control}
            errors={errors as FieldErrors}
            fieldArray={formFields}
            loading={false}
          />
        </div>
        <div css={styles.buttonsContainer}>
          <AppButton onClick={onCancel} variant="outlined" size="large">Cancel</AppButton>
          <AppButton
            style={styles.changePasswordBtn}
            onClick={handleSubmit(onSubmit)}
            disabled={!isValid}
            variant="contained"
            size="large"
          >
            Change password
          </AppButton>
        </div>
      </div>
    </BaseModal>
  );
};

export default ChangePasswordModal;
