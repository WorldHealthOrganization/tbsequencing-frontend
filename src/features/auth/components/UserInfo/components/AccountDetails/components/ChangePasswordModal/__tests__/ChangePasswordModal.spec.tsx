import React from 'react';
import { render, getTreeJSONWithRedux } from '../../../../../../../../../utils/testUtils/jestUtils';
import ChangePasswordModal from '../ChangePasswordModal';

describe('ChangePasswordModal component test', () => {
  test('renders ChangePasswordModal component without error', () => {
    render(
      <ChangePasswordModal />,
    );
  });

  it('Should render correctly', () => {
    const tree = getTreeJSONWithRedux(
      <ChangePasswordModal />,
    );

    expect(tree).toMatchSnapshot();
  });
});
