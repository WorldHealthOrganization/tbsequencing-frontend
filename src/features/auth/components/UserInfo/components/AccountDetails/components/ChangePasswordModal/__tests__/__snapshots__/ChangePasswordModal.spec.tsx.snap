// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangePasswordModal component test Should render correctly 1`] = `null`;
